pkg/csr/csr.go

```diff
@@ -0,0 +1,357 @@
+	return true, nil
+}
+
+// validateNodeName returns true if the node name passed here matches either the
+// actual host name of the VM'S or the reverse lookup of the instance addresses
+// present in the configMap. If the host name matches, it also validates if the
+// node name complies with the DNS RFC1123 naming convention.
+func (a *Approver) validateNodeName(nodeName string) (bool, error) {
+	configMap := &core.ConfigMap{}
+	err := a.client.Get(context.TODO(), kubeTypes.NamespacedName{Namespace: a.namespace,
+		Name: parser.InstanceConfigMap}, configMap)
+	if err != nil && !k8sapierrors.IsNotFound(err) {
+		return false, errors.Wrapf(err, "could not retrieve Windows instance configMap %s",
+			parser.InstanceConfigMap)
+	}
+
+	nodes := &core.NodeList{}
+	if err := a.client.List(context.TODO(), nodes); err != nil {
+		return false, errors.Wrap(err, "error listing nodes")
+	}
+
+	// Get the list of instances that are expected to be Nodes
+	windowsInstances, err := parser.ParseInstances(configMap.Data, nodes)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to parse hosts from ConfigMap %s", configMap.Name)
+	}
+	// check if the node name matches the lookup of any of the instance addresses
+	hasEntry, err := hasAssociatedInstance(windowsInstances, nodeName)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to map node name to the instances present in the configMap %s",
+			configMap.Name)
+	}
+	if hasEntry {
+		return true, nil
+	}
+	// Create a new signer using the private key secret
+	instanceSigner, err := signer.Create(kubeTypes.NamespacedName{Namespace: a.namespace,
+		Name: secrets.PrivateKeySecret}, a.client)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to create signer from private key secret")
+	}
+	// check if the node name matches any of the instance's host names
+	hasEntry, err = hasAssociatedHost(nodeName, windowsInstances, instanceSigner)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to map node name to the host names of instances in configMap %s",
+			configMap.Name)
+	}
+	if !hasEntry {
+		return false, nil
+	}
+	// validate node name for DNS RFC1123 naming conventions
+	// ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
+	if errs := validation.IsDNS1123Subdomain(nodeName); len(errs) > 0 {
+		a.recorder.Eventf(a.csr, core.EventTypeWarning, "nodeNameValidationFailed",
+			"node name %s does not comply with naming rules defined in RFC1123: "+
+				"Requirements for internet hosts", nodeName)
+		return false, nil
+	}
+	return true, nil
+}
+
+// validateKubeletServingCSR validates a kubelet serving CSR for its contents
+func (a *Approver) validateKubeletServingCSR(parsedCsr *x509.CertificateRequest) (bool, error) {
+	if a.csr == nil || parsedCsr == nil {
+		return false, errors.New("CSR or request should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(a.csr.Spec.Groups) < 2 {
```

Comment:
Is this check enough or should this iterate over the groups to the right values exist? Could be a helper similar to `hasUsages`